{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "97537554-152e-4639-8283-3007feb980a8",
   "metadata": {},
   "source": [
    "# Computational analysis of spatial transcriptomics data\n",
    "## BrainOmics 2.0 - Day 3\n",
    "\n",
    "#### 20th November 2024\n",
    "#### Ahmed Mahfouz, Benedetta Manzato (LUMC) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67603fb8-9267-4074-a900-f8e72f97cf94",
   "metadata": {},
   "source": [
    "### Agenda:\n",
    "<font size=\"3\">1. Spatial data download and format\n",
    "\n",
    "<font size=\"3\">2. Spatial imputation with SpaGE and Tangram\n",
    "\n",
    "<font size=\"3\">3. Spatial clustering with BANKSY"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e75a8e87-679c-42b7-8228-cdac0d35a39e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Dataset: MERFISH spatial transcriptomics dataset of a single adult mouse brain (Zhuang-ABCA-1) (Xiaowei Zhuang)\n",
    "\n",
    "<font size=\"3\"> This spatially resolved cell atlas of the whole mouse brain provides a systematic characterization of the spatial organization of transcriptomically defined cell types across the entire adult mouses brain using in situ, single-cell transcriptomic profiling with multiplexed error-robust fluorescence in situ hybridization (MERFISH). A gene panel of 1,122 genes was imaged in 9 million brain cells using **MERFISH**, and spatially resolved, single-cell expression profiling was performed at the whole-transcriptome scale by integrating the MERFISH data with the whole-brain single-cell RNA-sequencing (scRNA-seq) dataset previously generated by the Allen Institute. This resulted in the generation of a comprehensive cell atlas of more than 5,000 transcriptionally distinct cell clusters, belonging to âˆ¼300 major cell types (subclasses), across the whole mouse brain, with high molecular and spatial resolution.\n",
    "The authors performed MERFISH imaging on **245 coronal and sagittal sections**, the cell atlas was then spatially registered to the Allen Mouse Brain Common Coordinate Framework (CCFv3), which allows systematic quantifications of the cell composition and organization in individual brain regions as anatomically delineated in the CCFv3. This cell atlas also reveals molecularly-defined brain regions characterized by distinct cell type compositions, spatial gradients featuring gradual changes in the gene-expression profiles of cells, as well as cell-type-specific cell-cell interactions and the molecular basis and functional implications of these cell-cell interactions.\n",
    "\n",
    "<font size=\"3\"> The collection spans four mouse specimens (two coronal sets and two sagittal sets):\n",
    "- <font size=\"3\">**Zhuang-ABCA-1**: \n",
    "  - **Dataset size**: 4.2 million cells spanning 147 coronal sections with a 1,122 gene panel.\n",
    "  - **Filtered dataset**: 2.8 million cells passed the cell classification confidence score threshold and are displayed in the ABC atlas.\n",
    "  \n",
    "- <font size=\"3\">**Zhuang-ABCA-2**: \n",
    "  - **Dataset size**: 1.9 million cells spanning 66 coronal sections.\n",
    "  - **Filtered dataset**: 1.2 million cells passed the cell classification confidence score threshold.\n",
    "  \n",
    "- <font size=\"3\">**Zhuang-ABCA-3**: \n",
    "  - **Dataset size**: 2.1 million cells spanning 23 sagittal sections.\n",
    "  - **Filtered dataset**: 1.6 million cells passed the confidence score threshold.\n",
    "\n",
    "- <font size=\"3\">**Zhuang-ABCA-4**: \n",
    "  - **Dataset size**: 0.22 million cells spanning 3 sagittal sections.\n",
    "  - **Filtered dataset**: 0.16 million cells passed the confidence score threshold.\n",
    "\n",
    "\n",
    "<font size=\"3\"> We will work with the Brain Coronal 1 collection (Zhuang-ABCA-1); the cell-by-gene matrix of the 8.4 millions cells can be downloaded from the AWS bucket of this animal. The cells are then filtered by cell-classification (label transfer) confidence scores calculated during MERFISH-scRNAseq data integration. 7.0 million cells passed the confidence score threshold for cell subclass label transfer and 5.8 million cells further passed the confidence score threshold for cell cluster label transfer. These 5.8 million cells are included in the cell metadata file that can be downloaded from the the AWS bucket and are displayed on the ABC Atlas. The CCF coordinates of the 5.4 million cells that were registered to the 3D Allen-CCF can be downloaded from the CCF coordinate files in the AWB bucket. The collection spans four mouse specimens (2 coronal sets and 2 sagittal sets). Cells are mapped to the whole mouse brain taxonomy (WMB-taxonomy) and Allen Common Coordinate Framework (Allen-CCF-2020). \n",
    "Refer to Zhang et al, 2023 for more details.\n",
    "\n",
    "\n",
    "For more details: [Molecularly defined and spatially resolved cell atlas of the whole mouse brain (Zhang et al. 2023)](https://www.nature.com/articles/s41586-023-06808-9)\n",
    "\n",
    "From https://github.com/AllenInstitute/abc_atlas_access/blob/main/notebooks/zhuang_merfish_tutorial.ipynb\n",
    "\n",
    "Data Availablity : https://knowledge.brain-map.org/data/5C0201JSVE04WY6DMVC/collections </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13a8305d-204c-4780-8c01-738bbe0eeabd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "# Record the start time\n",
    "start_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c80f5d54-37d4-4375-96c9-48f29a1585a9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# Importing necessary libraries for spatial data analysis and visualization\n",
    "import os, re\n",
    "import scanpy as sc\n",
    "import scanpy.external as sce\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import random\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "\n",
    "# Display the version of important packages for reproducibility\n",
    "sc.logging.print_header()\n",
    "\n",
    "# Set the default aesthetics for Scanpy plots\n",
    "sc.set_figure_params(facecolor='white', figsize=(8, 8))  # White background, figure size 8x8\n",
    "sc.settings.verbosity = 0  # Set verbosity to show only errors (0) # errors (0), warnings (1), info (2), hints (3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f01cbf37-960e-46b1-b0e4-e05d32667014",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Importing functions to handle file paths and download data from ABC Atlas\n",
    "from pathlib import Path  # Object-oriented file paths\n",
    "from abc_atlas_access.abc_atlas_cache.abc_project_cache import AbcProjectCache  # Cache system for ABC Atlas data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d917233-09ff-4c30-8941-30a424c4f666",
   "metadata": {},
   "source": [
    "<font size=\"3\"> We create an output folder in your local space to save outputs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "860e1d4a-b95e-4d5f-84e3-ffee578f8f87",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder_path = './outputs/'\n",
    "if not os.path.exists(folder_path):\n",
    "        os.makedirs(folder_path)\n",
    "        print(f\"Folder created: {folder_path}\")\n",
    "else:\n",
    "    print(f\"Folder already exists: {folder_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6f83355-5d90-48d8-b358-e1b7ffc284d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "spatial_wd = '/group/brainomics/InputData'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cb55ab1-4073-47ac-87da-131d8d1d840f",
   "metadata": {},
   "source": [
    "<font size=\"3\">We will use [AbcProjectCache](https://github.com/AllenInstitute/abc_atlas_access/tree/41eb836e41e516ee528c98bf3979d0aff60c0b85) to download the data required for the tutorials.</font>\n",
    "\n",
    "<font size=\"3\">Data associated with the Allen Brain Cell Atlas is hosted on Amazon Web Services (AWS) in an S3 bucket as a AWS Public Dataset. No account or login is required. The S3 bucket is located here arn:aws:s3:::allen-brain-cell-atlas. You will need to be connected to the internet to run this notebook.\n",
    "\n",
    "<font size=\"3\">Each data release has an associated manifest.json which lists all the specific version of directories and files that are part of the release.</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2979380-6faf-4c10-81b7-843e208ea699",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Define the base path where data will be downloaded\n",
    "download_base = Path('./outputs/abc_download_root')  # Path to save downloaded data\n",
    "\n",
    "# Initialize the ABC Project Cache to manage downloaded data from the ABC Atlas\n",
    "abc_cache = AbcProjectCache.from_s3_cache(download_base)  # Fetch data from S3 cache\n",
    "\n",
    "# abc_cache.current_manifest show the current cache manifest\n",
    "abc_cache.current_manifest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d8506bf-adc6-4c38-ab16-c0d5b35d81fc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "abc_cache.load_latest_manifest()\n",
    "\n",
    "# List available directories\n",
    "abc_cache.list_directories"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9968f120-9669-4d22-ba9d-c30511af5771",
   "metadata": {},
   "source": [
    "<font size=\"3\">We will use the **Zhuang-ABCA-1 dataset**: 4.2 million cell spatial transcriptomics dataset spanning 147 coronal sections with a 1122 gene panel. 2.8 million cells passed cell classification confidence score threshold and displayed in the ABC atlas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e626af2-7515-47e1-b806-fd6c3ff4f8be",
   "metadata": {},
   "source": [
    "<font size=\"3\">List data files available in the Zhuang-ABCA-1 directory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "698f03e8-50c5-4854-8603-86888aac683a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "abc_cache.list_data_files('Zhuang-ABCA-1') "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2be22c8a-69ec-4ecf-ac54-7c608f6a8bdf",
   "metadata": {},
   "source": [
    "<font size=\"3\">List metadata files available in the Zhuang-ABCA-1 directory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5178bd4-2533-448c-a4c2-6622e4d3846c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "abc_cache.list_metadata_files('Zhuang-ABCA-1')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a775c5b-9e20-463a-9018-c6750adf1058",
   "metadata": {},
   "source": [
    "<font size=\"3\">Check the size of data and metadata:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59b39582-9e06-4269-a825-4c3c540a5eec",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "print(f\"Size of data: {abc_cache.get_directory_data_size('Zhuang-ABCA-1')}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "412e6b03-1ec9-42f4-894d-a26e6757890c",
   "metadata": {},
   "source": [
    "<font size=\"3\">Obtain path of data and metadata:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec0b9b91-4460-48a0-a267-3e18dba97a86",
   "metadata": {
    "tags": []
   },
   "source": [
    "```python\n",
    "path_data = abc_cache.get_directory_data('Zhuang-ABCA-1')\n",
    "print(\"Zhuang-ABCA-1 data files:\\n\\t\", path_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c87ce1e-a30d-4141-92bd-e94b5b7dcdb8",
   "metadata": {
    "tags": []
   },
   "source": [
    "```python\n",
    "path_metadata = abc_cache.get_directory_metadata('Zhuang-ABCA-1')\n",
    "print(\"Zhuang-ABCA-1 metadata files:\\n\\t\", path_metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c6e8277-93bb-4f68-8cc8-91d816393e53",
   "metadata": {},
   "source": [
    "<font size=\"3\">We previously downloaded the data for you, you can find them in `abc_download_root/`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9422d000-2e27-4c09-b562-49b2b4060952",
   "metadata": {},
   "source": [
    "### Data format\n",
    "\n",
    "<font size=\"3\"> Expression matrices are stored in the [anndata h5ad format](https://anndata.readthedocs.io/en/latest/) and need to be downloaded to a local file system for usage.\n",
    "\n",
    "<font size=\"3\">**Key Components of an AnnData Object:**\n",
    "\n",
    "<font size=\"3\">- **`.X`**: The main data matrix (usually cells by genes), where the rows are cells and the columns are genes. It typically stores raw or normalized gene expression data.\n",
    "\n",
    "<font size=\"3\">- **`.obs`**: Observations (metadata about the cells), such as cell type, experimental conditions, or quality control metrics.\n",
    "\n",
    "<font size=\"3\">- **`.var`**: Variables (metadata about the genes), such as gene names or other annotations.\n",
    "\n",
    "<font size=\"3\">- **`.obsm`**: Multi-dimensional annotations of cells, such as dimensionality reduction results (e.g., PCA, UMAP coordinates).\n",
    "\n",
    "<font size=\"3\">- **`.uns`**: Unstructured annotations, typically storing experimental metadata or plot settings.\n",
    "\n",
    "<font size=\"3\">- **`.layers`**: Stores multiple versions of the data matrix (e.g., raw counts, normalized data, etc.).\n",
    "\n",
    "<figure>\n",
    "    <img src=\"img/anndata_format.png\" alt=\"AnnData format\" width=\"500\" alt=\"Description of the image\">\n",
    "    <figcaption>From <a href=\"https://joss.theoj.org/papers/10.21105/joss.04371\" target=\"_blank\">Virshup et al. (2024) JOSS</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "\n",
    "<font size=\"3\">Spatial coordinates, being metadata about cells/spots are stored either in **`.obs`** or **`.obsm`**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e2fecd0-8ebd-4e6a-ae57-2ddc95588372",
   "metadata": {
    "tags": []
   },
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **Read raw count AnnData object first:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "968b9679-277b-44d7-890e-72d8369987bf",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## EX 1\n",
    "### Hint: use the function ad.read_h5ad \n",
    "\n",
    "path_raw_adata = f\"{spatial_wd}/day3/abc_download_root/expression_matrices/Zhuang-ABCA-1/Zhuang-ABCA-1-raw.h5ad\"\n",
    "\n",
    "# don't change the variable name\n",
    "adata = "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce8f7acd-bf12-40fb-bc80-b90528b137b1",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **Now read log-transformed AnnData object**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86a53987-1024-4801-bbdb-55fb698df137",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 2\n",
    "### Hint: use the function ad.read_h5ad \n",
    "\n",
    "path_log_adata = f\"{spatial_wd}/day3/abc_download_root/expression_matrices/Zhuang-ABCA-1/Zhuang-ABCA-1-log2.h5ad\"\n",
    "\n",
    "# don't change the variable name\n",
    "adata_log = "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "387bf361-673e-4d54-ac86-07904b2a457f",
   "metadata": {},
   "source": [
    "<font size=\"3\">Now we need to add the log-transformed data to the `adata` object\n",
    "\n",
    "<span style=\"color: red; font-size: 15px;\"> **To which component should it be added to?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db9276fc-6738-49e5-9588-8f34db172ebc",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 3\n",
    "# Add the log-transformed counts to the 'log' layer and the count data to the 'raw' layer\n",
    "### Hint: what slot of the adata object is suitable for count matrices?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1649782a-9579-430b-9c5f-ee27a22f6a46",
   "metadata": {},
   "source": [
    "<font size=\"3\">We delete adata_log since it won't be used anymore:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d467863b-4f3f-4004-9570-27be3deea412",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54eefccf-e216-4706-aa13-1c3df6daa599",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **Read metadata:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9479ab6d-2cd9-4d92-a0b2-ed33f30b50de",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 4\n",
    "# Read genes\n",
    "### Hint: use the function pd.read_csv and keep the gene identifier as index\n",
    "path_metadata_genes = f\"{spatial_wd}/day3/abc_download_root/metadata/Zhuang-ABCA-1/gene.csv\"\n",
    "\n",
    "genes ="
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "936c1cb3-d107-483c-a97e-ae3f1b1434ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 5\n",
    "# Read cells\n",
    "### Hint: use the function pd.read_csv and keep the cell label as index\n",
    "path_metadata_cell = f\"{spatial_wd}/day3/abc_download_root/metadata/Zhuang-ABCA-1/views/cell_metadata_with_cluster_annotation.csv\"\n",
    "\n",
    "cell_metadata_extended ="
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f2fbe3e-179f-4b4d-8d3e-87f0666fcb01",
   "metadata": {},
   "source": [
    " <span style=\"color: red; font-size: 15px;\"> **Add gene names to the right component in the `adata` object**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e11ccc4-adfd-43d9-9395-dcccc2d5673b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 6\n",
    "## Where should we store the gene metadata?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b1e07d6-c70f-4458-83c7-1e79e08d53f9",
   "metadata": {},
   "source": [
    "<font size=\"3\">Let's show some stats about the dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58b18b6f-b176-4fc8-93e9-1722c0ce2d86",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Number of sections in the dataset: {len(cell_metadata_extended['brain_section_label'].unique())}\")\n",
    "print(f\"Total number of cells/spots in the dataset: {cell_metadata_extended.shape[0]}\")\n",
    "print(f\"Average number of cells/spots per section: {cell_metadata_extended.shape[0] / len(cell_metadata_extended['brain_section_label'].unique())}\")\n",
    "print(f\"Size of the gene panel: {adata.shape[1]}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a200654d-4f37-43ad-ad71-8bcf2809df59",
   "metadata": {},
   "source": [
    "<font size=\"3\">We decided to work with **one section only**; let's select the one that captures most cells\n",
    "\n",
    "\n",
    "<img src=\"img/ABCAtlas_Zhuang-ABCA-1_Coronal Grid.png\" width=\"800\" height=\"300\">\n",
    "\n",
    "<span style=\"color: red; font-size: 15px;\"> **Which section has captures the highest number of cells?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2001f8c-4aa9-49e9-bfd7-98dc74f26a6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 7\n",
    "### Hint : use collections.Counter - https://docs.python.org/3/library/collections.html#collections.Counter\n",
    "import collections \n",
    "\n",
    "### Steps: \n",
    "# 1. Identify the brain_section_label of the section with the highest number of cells\n",
    "# 2. Subset cell_metadata_extended and adata to include only cells from that section\n",
    "\n",
    "# Output subset info\n",
    "print(f\"Subset contains {adata_section.shape[0]} cells from section {top_section_label}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c2de437-f7b2-407b-b950-9e098db110c3",
   "metadata": {},
   "source": [
    "<font size=\"3\">Let's inspect `adata_section` and `cell_metadata_extended_sub`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30971d4e-e230-4fa6-9667-d9212b774adf",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "48044fd0-212e-4fd1-9cd5-aa4a1efc846b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_metadata_extended_sub.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "338c1aca-e8b5-4d17-827f-aa7eb04d442a",
   "metadata": {},
   "source": [
    "<font size=\"3\"> We notice that the indexes don't match! It's important to reorder the indexes in cell_metadata_extended_sub to match the ones in adata_section.obs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "927535a0-b816-414c-a015-27c9518fba56",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_metadata_extended_sub_reind = cell_metadata_extended_sub.reindex(adata_section.obs.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a62ea072-2d64-4622-81e7-e1958be2d164",
   "metadata": {},
   "outputs": [],
   "source": [
    "collections.Counter(cell_metadata_extended_sub_reind.index == adata_section.obs.index)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "48917161-9707-45f9-acf3-d6ff056295fb",
   "metadata": {},
   "source": [
    "<font size=\"3\">Now we can add cell_metadata_extended_sub as metadata to adata_section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "febd57ff-bce7-42b1-811d-19179435e221",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section.obs = cell_metadata_extended_sub_reind"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2edc2f71-e0b9-4ad4-8bcd-4cab02b4b939",
   "metadata": {},
   "source": [
    "<font size=\"3\">Plot the section and color them by all the different classifications avaiable in the metadata (class_color, subclass_color, supertype_color, cluster_color):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8470a4f6-e392-42f3-b15f-2a3f95011163",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the color mapping dictionaries\n",
    "class_palette = dict(zip(adata_section.obs['class'].unique(), adata_section.obs['class_color'].unique()))\n",
    "subclass_palette = dict(zip(adata_section.obs['subclass'].unique(), adata_section.obs['subclass_color'].unique()))\n",
    "supertype_palette = dict(zip(adata_section.obs['supertype'].unique(), adata_section.obs['supertype_color'].unique()))\n",
    "cluster_palette = dict(zip(adata_section.obs['cluster'].unique(), adata_section.obs['cluster_color'].unique()))\n",
    "\n",
    "# Create the subplots\n",
    "fig, axes = plt.subplots(2, 2, figsize=(15, 15))\n",
    "fig.suptitle('Scatter Plots of Different Categories', fontsize=16)\n",
    "\n",
    "# Plot 'class' using class_color\n",
    "sns.scatterplot(ax=axes[0, 0], x='x', y=-adata_section.obs['y'], hue='class', palette=class_palette, data=adata_section.obs, s=8, legend=False)\n",
    "axes[0, 0].set_title('Class')\n",
    "axes[0, 0].set_xlabel('X Coordinate')\n",
    "axes[0, 0].set_ylabel('Y Coordinate')\n",
    "axes[0, 0].tick_params(axis='both', which='both', bottom=False, left=False, labelbottom=False, labelleft=False)\n",
    "axes[0, 0].grid(False)  # Remove grid\n",
    "\n",
    "# Plot 'subclass' using subclass_color\n",
    "sns.scatterplot(ax=axes[0, 1], x='x', y=-adata_section.obs['y'], hue='subclass', palette=subclass_palette, data=adata_section.obs, s=8, legend=False)\n",
    "axes[0, 1].set_title('Subclass')\n",
    "axes[0, 1].set_xlabel('X Coordinate')\n",
    "axes[0, 1].set_ylabel('Y Coordinate')\n",
    "axes[0, 1].tick_params(axis='both', which='both', bottom=False, left=False, labelbottom=False, labelleft=False)\n",
    "axes[0, 1].grid(False)  # Remove grid\n",
    "\n",
    "# Plot 'supertype' using supertype_color\n",
    "sns.scatterplot(ax=axes[1, 0], x='x', y=-adata_section.obs['y'], hue='supertype', palette=supertype_palette, data=adata_section.obs, s=8, legend=False)\n",
    "axes[1, 0].set_title('Supertype')\n",
    "axes[1, 0].set_xlabel('X Coordinate')\n",
    "axes[1, 0].set_ylabel('Y Coordinate')\n",
    "axes[1, 0].tick_params(axis='both', which='both', bottom=False, left=False, labelbottom=False, labelleft=False)\n",
    "axes[1, 0].grid(False)  # Remove grid\n",
    "\n",
    "# Plot 'cluster' using cluster_color\n",
    "sns.scatterplot(ax=axes[1, 1], x='x', y=-adata_section.obs['y'], hue='cluster', palette=cluster_palette, data=adata_section.obs, s=8, legend=False)\n",
    "axes[1, 1].set_title('Cluster')\n",
    "axes[1, 1].set_xlabel('X Coordinate')\n",
    "axes[1, 1].set_ylabel('Y Coordinate')\n",
    "axes[1, 1].tick_params(axis='both', which='both', bottom=False, left=False, labelbottom=False, labelleft=False)\n",
    "axes[1, 1].grid(False)  # Remove grid\n",
    "\n",
    "plt.tight_layout(rect=[0, 0, 1, 0.96])  \n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "246a60af-ec8e-4beb-9d43-722916f62aec",
   "metadata": {},
   "source": [
    "### Quality Control\n",
    "\n",
    "<font size=\"3\">From https://scanpy-tutorials.readthedocs.io/en/latest/spatial/basic-analysis.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bbfbbab-1a81-4092-92df-445891b6b179",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make gene names unique\n",
    "adata_section.var_names_make_unique()\n",
    "# add mt percentage\n",
    "adata_section.var[\"mt\"] = adata_section.var_names.str.startswith(\"MT-\")\n",
    "sc.pp.calculate_qc_metrics(adata_section, qc_vars=[\"mt\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16284989-0551-4c98-b12e-f909aece2eb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "sns.histplot(adata_section.obs[\"total_counts\"], kde=False, ax=axs[0])\n",
    "sns.histplot(\n",
    "    adata_section.obs[\"total_counts\"][adata_section.obs[\"total_counts\"] < 10000],\n",
    "    kde=False,\n",
    "    bins=40,\n",
    "    ax=axs[1],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94c92c43-6c7c-4712-af33-05c87f06845d",
   "metadata": {},
   "source": [
    "#### Cell Filtering and Preprocessing in Single-Cell RNA-Seq Analysis\n",
    "\n",
    "<font size=\"3\">In this section, we are performing a series of filtering steps to preprocess the single-cell RNA sequencing (scRNA-seq) data stored in the `adata_section` object. The filtering aims to retain high-quality cells and genes for downstream analyses. The following steps are applied:\n",
    "\n",
    "1. <font size=\"3\">**Filter Cells by Minimum Counts**: \n",
    "   - <font size=\"3\">We filter out cells that have fewer than 200 counts. This helps to exclude low-quality cells that may not provide reliable information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e2c147d1-483f-4ee4-85ba-ed310739df71",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(adata_section, min_counts=200)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "250398ee-bce6-4ff0-8785-d37723b3b05b",
   "metadata": {},
   "source": [
    "2. <font size=\"3\">**Filter Cells by Maximum Counts**: \n",
    "   - <font size=\"3\">We also filter out cells with more than 25,000 counts to remove potential doublets or outlier cells that may artificially inflate the count data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67cc3092-4379-4b6b-875d-38edbe3d17bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(adata_section, max_counts=25000)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d5f5af3-5eee-492a-8f15-4d693e9ef6dc",
   "metadata": {},
   "source": [
    "3. <font size=\"3\">**Filter Based on Mitochondrial Gene Expression**:\n",
    "\n",
    "   - <font size=\"3\">We further refine our dataset by excluding cells that have more than 10% of their total counts coming from mitochondrial genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43186dec-2f23-4b81-8ebf-2892979a2d70",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section = adata_section[adata_section.obs[\"pct_counts_mt\"] < 10].copy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac747ef5-f495-4346-84cd-bb9d6dff3bbc",
   "metadata": {},
   "source": [
    "4. <font size=\"3\">**Filter Genes by Minimum Cells**:\n",
    "\n",
    "- <font size=\"3\">Finally, we filter out genes that are detected in fewer than 20 cells, ensuring that we retain only those genes that have sufficient expression across our dataset."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34c8b911-1e97-4b42-be53-385753719293",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **Filter out genes that are detected in less than 20 cells**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ba80114-d852-42f6-b700-70b596160a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 8\n",
    "### Hint: use sc.pp.filter_genes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1da18a6-f45b-44ed-a4eb-59a8cc7a9902",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Number of cells after filtering: {adata_section.n_obs}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34f11a84-c472-47a7-a3b5-1ee3569b753c",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Perform PCA and UMAP and cluster with Leiden Algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77137142-5f01-452a-9015-7dc8b82dfd15",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.pca(adata_section)\n",
    "sc.pp.neighbors(adata_section)\n",
    "sc.tl.umap(adata_section)\n",
    "sc.tl.leiden(\n",
    "    adata_section, \n",
    "    key_added=\"clusters\", \n",
    "    resolution=1.0,\n",
    "    directed=False, \n",
    "    n_iterations=2\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "871dc765-9919-444e-ae31-f0c0be0b97c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (8,8)\n",
    "sc.pl.umap(adata_section, color=[\"total_counts\", \"n_genes_by_counts\", \"clusters\"], wspace=0.2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec10e64a-9e5f-47aa-83a2-2b3567c6f84f",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Some stats:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b5e5424-5188-49ed-9678-b048bb2c5121",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Number of Neurotransmitters: ', len(adata_section.obs['neurotransmitter'].unique()))\n",
    "print('Number of Classes: ', len(adata_section.obs['class'].unique()))\n",
    "print('Number of Subclasses: ', len(adata_section.obs['subclass'].unique()))\n",
    "print('Number of Supertype: ', len(adata_section.obs['supertype'].unique()))\n",
    "print('Number of Cluster: ', len(adata_section.obs['cluster'].unique()))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11662910-aee3-4a58-a087-a7136c77c3d1",
   "metadata": {},
   "source": [
    "<font size=\"3\"> The filtering step performed by the authors of the paper is done differently than the basic one we just performed. The metadata is given only for the filtered object. For this reason we will filter our object to keep only spots that passed the quality control for the Allen Brain Atlas standards."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2c9769e-7819-447b-a13c-968f377167ee",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# filter spots in adata_section and keep only the ones that are present in metadata\n",
    "adata_section = adata[adata.obs.index.isin(cell_metadata_extended_sub.index)]\n",
    "# reindex cell_metadata_ext_sub to match the index order of adata_section.obs\n",
    "cell_metadata_extended_sub = cell_metadata_extended_sub.reindex(adata_section.obs.index)\n",
    "# add metadata to the metadata field in an anndata object : adata.obj\n",
    "adata_section.obs = cell_metadata_extended_sub\n",
    "\n",
    "print(f\"The filtered Section 80 dataset has {adata_section.shape[0]} cells and {adata_section.shape[1]} genes.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69f0267e-3281-416f-9bb5-576651dee504",
   "metadata": {},
   "source": [
    "<font size=\"3\"> To free RAM, we delete the `adata` object as we will use only the section:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7528d536-10e2-4e98-baa9-a7d429deafe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d248bde-69b8-42a8-95e5-84c7fb2581f3",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Some spatial tools require the spatial coordinates ti be stored in `.obsm` as `spatial` or `X_spatial`\n",
    "\n",
    "<font size=\"3\"> This is how you can add them:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10c021bf-aa04-4477-a82c-9a9d6c06cdf8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add X_spatial\n",
    "adata_section.obsm['X_spatial'] = np.array(adata_section.obs[['x','y']])\n",
    "adata_section.obsm['X_spatial']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5bcff9f7-97b8-4664-acde-be61cabc2239",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Once the spatial coordinates are stored in  `.obsm[X_spatial]` you can plot the section with `sc.pl.spatial`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e307ded4-983e-48d4-8a7c-42c4f5f6e66a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the spatial plot\n",
    "sc.pl.spatial(adata_section, color='class', color_map='Accent_r', spot_size=0.04, title='Section 80 colored by Class')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eadb60f0-7cbb-45d6-8ad4-8567079ed879",
   "metadata": {},
   "source": [
    "### Plot gene expression in space\n",
    "\n",
    "<font size=\"3\">\n",
    "\n",
    "Since each spot in the spatial data corresponds to a specific location and its associated gene expression profile, we can visualize spatial gene expression.\n",
    "\n",
    "We can do this using the function `sc.pl.spatial`, it requires the name of the variable index to indicate the genes to plot.\n",
    "\n",
    "Let's take a look at `adata_section`.var:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d001c6ad-0581-4e54-bcc8-5ea0db2d147c",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b82b4f1-d8d8-44e6-97f2-81f2dfbee5ed",
   "metadata": {},
   "source": [
    "<font size=\"3\">\n",
    "\n",
    "Change the index of the `adata_section.var` to `gene_symbol` for clarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf8740b3-6142-42df-8318-1179a739d04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset the index to turn the current index into a column\n",
    "adata_section.var.reset_index(inplace=True)\n",
    "adata_section.var.set_index('gene_symbol', inplace=True)\n",
    "adata_section.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11c787c1-911c-4bbb-b621-fee7010bbeb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define genes to plot\n",
    "genes_toplot = ['Zic1',\n",
    " 'Penk',\n",
    " 'Gfap',\n",
    " 'Dlx6',\n",
    " 'Cux2',\n",
    " 'Fezf2',\n",
    " 'Reln',\n",
    " 'Nr4a2',\n",
    " 'Satb2']\n",
    "\n",
    "sc.pl.spatial(adata_section, color=genes_toplot,spot_size=0.04)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c2e379a-79f5-46ff-94a0-654f9026c150",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **Plot your favourite gene/s**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "643caa15-8d5d-472e-8668-97fd010b768e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genes_toplot = [.....]\n",
    "\n",
    "#sc.pl.spatial(adata_section, color=genes_toplot,spot_size=0.04)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bad7970-e5b8-421a-8e14-91c382d2eabc",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_section.write_h5ad(f\"./outputs/adata_section80.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c35bfe-bcf8-4d3b-8493-ca07c89f5c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Record the end time\n",
    "end_time = time.time()\n",
    "\n",
    "# Calculate elapsed time in minutes\n",
    "elapsed_time = (end_time - start_time) / 60\n",
    "print(f\"Execution Time: {elapsed_time:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42a3d2d5-5727-45df-a641-106b5e7c4460",
   "metadata": {},
   "source": [
    "# Prediction of spatial patterns for unmeasured genes using scRNA-seq data with SpaGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b24a6480-c29d-4399-97c5-1ceb21638f73",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f356ab26-9e1f-4c01-81e4-739efec8e1ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "from SpaGE.main import SpaGE\n",
    "import scipy.stats as st"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beef66b4-853d-403f-a573-52053b399a49",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <img src=\"img/spage.jpg\" width=\"700\" height=\"280\" alt=\"Description of the image\">\n",
    "    <figcaption>From <a href=\"https://academic.oup.com/nar/article/48/18/e107/5909530\" target=\"_blank\">Abdelaal et al. (2020) NAR</a></figcaption>\n",
    "</figure>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19eb7650-ac6d-4b90-aa06-6cc781c6311e",
   "metadata": {},
   "source": [
    "<font size=\"3\"> SpaGE takes as input the (targeted) spatial dataset (query) and a single-cell RNA dataset used as reference. The two datasets should be generated under the same conditions (same species, tissue, region) as much as possible.\n",
    "\n",
    "<font size=\"3\"> We will use [Mouse whole-brain transcriptomic cell type atlas (Hongkui Zeng)](https://www.nature.com/articles/s41586-023-06812-z).This scRNAseq dataset profiled 7 million cells (approximately 4.0 million cells passing quality control). More in detail:\n",
    "\n",
    " - <font size=\"3\"> 1.7 million single cell transcriptomes spanning the whole adult mouse brain using 10Xv2 chemistry (WMB-10Xv2)\n",
    "\n",
    " - <font size=\"3\"> 2.3 million single cell transcriptomes spanning the whole adult mouse brain using 10Xv3 chemistry (WMB-10Xv3)\n",
    "\n",
    " - <font size=\"3\"> 1687 single cell transcriptomes spanning the whole adult mouse brain using the 10X Multiome chemistry (WMB-10XMulti)\n",
    "\n",
    "<figure>\n",
    "    <img src=\"img/sc.png\" width=\"700\" height=\"280\" alt=\"Description of the image\">\n",
    "    <figcaption>From <a href=\"https://www.nature.com/articles/s41586-023-06812-z\" target=\"_blank\">Yao et al. (2023) Nature Methods</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "\n",
    "\n",
    "<font size=\"3\"> Let's check the names of the files in each of the three single-cell collections:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a26d4f58-a9de-461e-bdb6-23ff9a9f3307",
   "metadata": {},
   "outputs": [],
   "source": [
    "abc_cache.list_data_files('WMB-10Xv2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d51adb7-1297-434e-9b20-fa4fca8cdaf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "abc_cache.list_data_files('WMB-10Xv3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "822963e0-8edd-4141-825a-89964e4aff6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "abc_cache.list_data_files('WMB-10XMulti')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "127219b3-2b09-4e16-86ba-2cd1c229495c",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Let's check the size of the files in each of the three single-cell collections:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fbc4886-1bdf-46a2-8c35-6b491dc1e406",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Size of WMB-10Xv2 data: {abc_cache.get_directory_data_size('WMB-10Xv2')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a913a0a1-804b-4cb3-812a-9afe06a4ea5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Size of WMB-10Xv3 data: {abc_cache.get_directory_data_size('WMB-10Xv3')}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f5454ee-2afe-47cb-874d-9dc7e91f82d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Size of WMB-10X Multi data: {abc_cache.get_directory_data_size('WMB-10XMulti')}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3314c600-5aff-4998-9fde-c7dd6185a1dd",
   "metadata": {},
   "source": [
    "<font size=\"3\"> The size is of the combined single-cell object is around 280GB. For this reason, the three objects have been previously downloaded, concatenated, preprocessed and subsetted to obtain NNN cells. For your knowledge, you can find all the steps in `obtain_scref.py`.\n",
    "\n",
    "<font size=\"3\"> Let's read the resulting single-cell AnnData object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "926edba6-5733-4042-84f3-7443ed238f0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc = ad.read_h5ad(f\"{spatial_wd}/day3/SpaGE_data/sc_adata_1percent.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08a97c98-d1c9-42e0-8c2c-597b470d2580",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Inspect the object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e292f80-d9d6-4df6-b4c6-cf92723800d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02cec20e-26ee-4e40-a869-a79003d9df30",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Inspect the metadata:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "506a07d2-29c1-4a98-beba-45735f57efe8",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0e98222-9420-4181-868d-cd146b5396db",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Add feature metadata to the AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f6b28a7-7c33-446a-9294-aa92289833a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = pd.read_csv(f\"{spatial_wd}/day3/abc_download_root/metadata/WMB-10X/gene.csv\",index_col=0)\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc782a1c-66b6-4eb4-aa95-ea4a2eb6914f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc.var = genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5aec6d0c-dfa4-4586-b887-70330b219f35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reset the index to turn the current index into a column\n",
    "adata_sc.var.reset_index(inplace=True)\n",
    "adata_sc.var.set_index('gene_symbol', inplace=True)\n",
    "adata_sc.var.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4ac2775-5018-4eb9-a852-2d79a7ec95ce",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Set up inputs for SpaGE:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d55535e-9547-4d92-a173-81077340893b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the count data and transform them into a Pandas DataFrame\n",
    "RNA_data = pd.DataFrame(adata_sc.X.todense())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2eb576b9-394b-44c3-ad22-49a40e6c962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add gene names to the RNA_data columns \n",
    "RNA_data.columns = adata_sc.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3f266e6-82d4-4ced-8a1c-1aa5daa55ef2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter lowely expressed genes\n",
    "Genes_count = np.sum(RNA_data > 0, axis=1)\n",
    "RNA_data = RNA_data.loc[Genes_count >=10,:]\n",
    "del Genes_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecb28736-c54d-404f-b0e3-b4e09c99ac69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalization\n",
    "def Log_Norm_cpm(x):\n",
    "    return np.log(((x/np.sum(x))*1000000) + 1)\n",
    "RNA_data = RNA_data.apply(Log_Norm_cpm,axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77339674-5f9d-4ddb-b4ba-7cbf2278e417",
   "metadata": {},
   "outputs": [],
   "source": [
    "# transpose RNA_data to have rows = genes and columns = cells\n",
    "RNA_data = RNA_data.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6099b6ab-1c53-4ffa-be3c-c4cda0f5c69f",
   "metadata": {},
   "outputs": [],
   "source": [
    "RNA_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0ac960e-7c2c-471d-8d80-3bf4d82eee2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop any row with at least one NaN value\n",
    "RNA_data = RNA_data.dropna()  # This will drop any row with at least one NaN value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "119e8228-3eda-4786-8e83-a959d7757b44",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **How many genes do the spatial and single-cell data set have in common?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2846466a-d702-4e33-a9b3-8e6feafbdc5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 9\n",
    "## Hint: use .intersection\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e889dec-53c0-42bd-8d40-8e0c596f7661",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **How many genes are measured in the single-cell dataset that are not measured in the spatial dataset?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a774606-2b8e-4868-bda5-b4eeb8c13b42",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 10\n",
    "## Hint: use a set operation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d616ce55-41de-420b-afa3-93d50b8a461c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract count matrix from adata_section and rename index and column names\n",
    "\n",
    "spatial_data = pd.DataFrame(adata_section.X)\n",
    "spatial_data.index = adata_section.obs.index\n",
    "spatial_data.columns = adata_section.var.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a2608515-710a-462c-8ac2-e835eb413c6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalize spatial_data \n",
    "spatial_data = spatial_data.apply(Log_Norm_cpm,axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a7b24db-b2b6-4b53-861a-074d6edf7276",
   "metadata": {},
   "outputs": [],
   "source": [
    "brain_region_specific_genes = [\n",
    "    \"Bcl11b\", \"Cux2\", \"Fezf2\", \"Tbr1\", \"Satb2\", # isocortex\n",
    "    \"Nr4a2\",  # Midbrain specific\n",
    "    \"Foxp2\",  # Striatum specific\n",
    "    \"Penk\",   # Basal ganglia specific\n",
    "    \"Zic1\",   # Cerebellum specific\n",
    "    \"Hoxb2\",  # Hindbrain specific\n",
    "    \"Dlx6\",   # Subpallium specific\n",
    "]\n",
    "\n",
    "# Checking for overlap\n",
    "overlap_genes = list(set(list(spatial_data.columns)) & set(brain_region_specific_genes))\n",
    "overlap_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6720795-f4b1-4770-b965-60e510414ca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# non overlap genes\n",
    "non_overlap_genes = list(set(brain_region_specific_genes) - set(overlap_genes))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c1606da-6019-4423-ad49-8319dc69653d",
   "metadata": {},
   "source": [
    "### Leave-one-gene-out cross validation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88d6dbaa-dcbb-427c-97cd-5361a83b9840",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Define the function to predict genes measured in the spatial dataset to have a ground truth and plot the measured expression next to the predicted one.\n",
    "<font size=\"3\"> Calculate Spearman correlation between the measured and predicted expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b94062d-7ddc-40bc-9a49-7bbd1dfd9143",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.gridspec as gridspec  # Import for more precise subplot layout control\n",
    "plt.style.use('dark_background')\n",
    "\n",
    "def plot_measured_predicted(Gene_set, corr_df):\n",
    "\n",
    "    for i in Gene_set:\n",
    "        # Predict the expression of the current gene using the SpaGE method\n",
    "        Imp_Genes = SpaGE(spatial_data.drop(i, axis=1), RNA_data.T, n_pv=30, genes_to_predict=[i])\n",
    "    \n",
    "        # Calculate the Spearman correlation between measured and predicted values\n",
    "        corr_df[i] = st.spearmanr(spatial_data[i], Imp_Genes[i])[0]\n",
    "    \n",
    "        # Create a figure with two subplots (1 row, 2 columns)\n",
    "        fig = plt.figure(figsize=(15, 8))  # Adjust figure size if needed\n",
    "        gs = gridspec.GridSpec(1, 3, width_ratios=[1, 1, 0.05], wspace=0.3)  # Create a gridspec for better control\n",
    "    \n",
    "        # --- Measured Gene Expression Plot ---\n",
    "        ax0 = fig.add_subplot(gs[0])  # First plot\n",
    "        cmap_measured = spatial_data[i]\n",
    "        cmap_measured[cmap_measured > np.percentile(cmap_measured, 99)] = np.percentile(cmap_measured, 99)\n",
    "        sc1 = ax0.scatter(adata_section.obs['x'], -adata_section.obs['y'], s=1, c=cmap_measured, alpha=1)\n",
    "    \n",
    "        ax0.set_title(f'Measured Expression: {i}', fontsize=12)\n",
    "        ax0.set_xlabel('X coordinate')\n",
    "        ax0.set_ylabel('Y coordinate')\n",
    "        ax0.set_aspect('equal', adjustable='box')\n",
    "        ax0.grid(False)\n",
    "    \n",
    "        # --- Predicted Gene Expression Plot ---\n",
    "        ax1 = fig.add_subplot(gs[1])  # Second plot\n",
    "        cmap_predicted = Imp_Genes[i]\n",
    "        cmap_predicted[cmap_predicted > np.percentile(cmap_predicted, 99)] = np.percentile(cmap_predicted, 99)\n",
    "        sc2 = ax1.scatter(adata_section.obs['x'], -adata_section.obs['y'], s=1, c=cmap_predicted, alpha=1)\n",
    "    \n",
    "        ax1.set_title(f'Predicted Expression: {i}', fontsize=12)\n",
    "        ax1.set_xlabel('X coordinate')\n",
    "        ax1.set_ylabel('Y coordinate')\n",
    "        ax1.set_aspect('equal', adjustable='box')\n",
    "        ax1.grid(False)\n",
    "\n",
    "        # Create a colorbar on the far right\n",
    "        cbar_ax = fig.add_subplot(gs[2])  # Third subplot (narrow space for the color bar)\n",
    "        cbar = fig.colorbar(sc1, cax=cbar_ax)  # Assign the color bar to the third subplot\n",
    "        cbar.set_label(i)\n",
    "\n",
    "        # Show the plot\n",
    "        plt.tight_layout()  # Adjust spacing to avoid overlap\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95639806-6b66-4151-b829-de3c5a400507",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Since we have a ground truth, create a Series to store Spearman correlation results for each gene\n",
    "# Our gene_set in this case is a set of genes shared between the spatial and the sc datasets\n",
    "correlations_spage = pd.Series(index=overlap_genes)\n",
    "\n",
    "plot_measured_predicted(overlap_genes,correlations_spage)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6778c0bb-5b1f-41f7-b764-1ac748d2bed3",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Let's take a look at the Spearman Correlations between measured and predicted gene expression:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49dafbec-7b73-4ffc-ad3d-eb1576393a3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "correlations_spage = pd.DataFrame(correlations_spage, columns=[\"correlations_spage\"])\n",
    "correlations_spage"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb8d1594-6107-4869-9b38-d6e6b004845f",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **Is there any other gene expression (measured in the spatial dataset) you would like to predict?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9f32ba9-5e68-4492-9b57-7f7b1af954fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(gene_intersection)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f894fac-7ead-4104-95d7-6357cf2d1fe9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the set of genes (from the gene_intersection list) where we will perform leave-one-out cross-validation\n",
    "Gene_set = ['Grm3'] # has to be a list\n",
    "# Since we have a ground truth, create a Series to store Spearman correlation results for each gene\n",
    "correlations_spage_extragenes = pd.Series(index=Gene_set)\n",
    "\n",
    "plot_measured_predicted(Gene_set,correlations_spage_extragenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c389debc-7913-4833-b72a-fb7717edff50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print Correlations\n",
    "correlations_spage_extragenes = pd.DataFrame(correlations_spage_extragenes, columns=[\"correlations_spage\"])\n",
    "correlations_spage_extragenes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9bab07e-c287-4be4-bed0-dbfde2f9729e",
   "metadata": {},
   "source": [
    "### Predict unmeasured genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39d5347c-7c38-4c90-be33-08cb8262939c",
   "metadata": {},
   "source": [
    "<font size=\"3\">  Define a function to predict unmeasured genes in the spatial dataset. \n",
    "\n",
    "<font size=\"3\">  In this case correlations cannot be calculated as there is no ground truth available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5b9df78-120c-43d3-b850-c77537271492",
   "metadata": {},
   "outputs": [],
   "source": [
    "def predict_unmeasured_genes(Gene_set):\n",
    "    for i in Gene_set:\n",
    "    \n",
    "        # Predict the expression of the current gene using the SpaGE method\n",
    "        Imp_Genes = SpaGE(spatial_data, RNA_data.T, n_pv=30, genes_to_predict=[i])\n",
    "   \n",
    "        # Create a new figure for each gene's predicted expression\n",
    "        plt.figure(figsize=(10, 8))\n",
    "    \n",
    "        # Get predicted gene expression values and clip both lower and upper bounds\n",
    "        cmap_predicted = Imp_Genes[i]\n",
    "    \n",
    "        # Clip the minimum values at 0 (no negatives) and limit extreme values at 99th percentile\n",
    "        cmap_predicted = np.clip(cmap_predicted, 0, np.percentile(cmap_predicted, 99))\n",
    "    \n",
    "        # Plot scatter plot for the predicted gene expression\n",
    "        sc2 = plt.scatter(adata_section.obs['x'], -adata_section.obs['y'], s=1, c=cmap_predicted, alpha=1)\n",
    "    \n",
    "        # Set plot title and labels\n",
    "        plt.title(f'Predicted Expression: {i}', fontsize=12)\n",
    "        plt.xlabel('X coordinate')\n",
    "        plt.ylabel('Y coordinate')\n",
    "        plt.gca().set_aspect('equal', adjustable='box')\n",
    "        plt.grid(False)\n",
    "    \n",
    "        # Create a colorbar on the same plot\n",
    "        cbar = plt.colorbar(sc2)\n",
    "        cbar.set_label(i)\n",
    "    \n",
    "        # Adjust layout and display the plot\n",
    "        plt.tight_layout()\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab92cc4a-9eb4-4a73-b54c-cb8cb86edd2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the set of genes (from the non_overlap_genes list) to be predicted by SpaGE\n",
    "predict_unmeasured_genes(non_overlap_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06189e2c-8e20-4a83-8b5f-f231dc51b0cc",
   "metadata": {},
   "source": [
    "<font size=\"3\"> There is no definitive ground truth for the real gene expression profiles of these genes. However, an alternative approach could be to plot region-specific genes and verify whether their expression aligns with the expected brain regions.\n",
    "\n",
    "<font size=\"3\"> For example, Tbr1 is known to be Isocortex-specific; does its expression show higher levels in the Isocortex? Similarly, Foxp2 is Striatum-specific; does it exhibit a Striatum-specific expression pattern?\n",
    "\n",
    "<span style=\"color: red; font-size: 15px;\"> **Use [this](https://mouse.brain-map.org/search/index) resource to check region-specific gene expression.**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6360e33e-80cd-4e70-8835-040cefee7b07",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **Is there any other gene expression (NOT measured in the spatial dataset) you would like to predict?**\n",
    "\n",
    "<span style=\"color: red; font-size: 15px;\"> **Can you think of other genes that we could use to visually validate the accuracy of SpaGE by checking if their expression patterns correspond to their known brain regions? Plot them!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d62f84d-1144-44f7-8701-55fe5a99bd09",
   "metadata": {},
   "outputs": [],
   "source": [
    "#list(genes_not_intersection)[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5789f404-810a-4623-8f8f-c0d9b8c5be46",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Gene_set = [...]\n",
    "\n",
    "#predict_unmeasured_genes(Gene_set)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "371ac2df-aef5-493a-8ff5-38fb8aa29564",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Record the end time\n",
    "end_time = time.time()\n",
    "\n",
    "# Calculate elapsed time in minutes\n",
    "elapsed_time = (end_time - start_time) / 60\n",
    "print(f\"Execution Time SpaGE: {elapsed_time:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "948f6c80-567e-4f56-970d-00d8d1bdaf52",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Let's delete some objects we won't be using anymore to free some RAM:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e35c884-a934-4070-a0cd-c5d556db8d33",
   "metadata": {},
   "outputs": [],
   "source": [
    "del RNA_data, spatial_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "371a83c3-c3e8-4b01-a94e-9cd30c952ea3",
   "metadata": {},
   "source": [
    "# Unmeasured gene imputation with Tangram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "518b2c69-c6c0-4ba9-ada7-f7a1709c594b",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3000981c-012e-404e-b06e-409e7eaaf9cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import tangram as tg\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.preprocessing import MinMaxScaler\n",
    "\n",
    "sc.settings.verbosity = 3\n",
    "sc.settings.set_figure_params(dpi=80, facecolor=\"white\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92e8bf4e-ba3e-4559-8b22-fd89efaa0657",
   "metadata": {},
   "source": [
    "#### Common gene set between reference and spatial dataset\n",
    "\n",
    "<font size=\"3\"> We want to select our training genes. These genes are shared between the two datasets and should capture the biological variance between cell types. For this, we first compute marker genes on the single-cell data and then use Tangramâ€™s preprocessing function to subset to those genes that are also present in the spatial data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "674dbcef-8913-42f0-812a-50d342d679f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "markers = list(set.intersection(set(adata_sc.var_names), set(adata_section.var_names)))\n",
    "len(markers)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1740106c-6c6c-4dbb-b9d0-9cb5bd85ceb0",
   "metadata": {},
   "source": [
    "<font size=\"3\"> `tg.pp_adatas` computes the overlap between single-cell data and spatial data on the list of genes provided in the genes argument, it then stores the resulting gene set under 'training_genes' in both adata objects under the `.uns` key and enforces consistent ordering of the genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d12abfa2-3533-4ac3-a16a-55357663742a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tg.pp_adatas(adata_sc, adata_section, genes=markers) # using the same single-cell reference that we used to run SpaGE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fc9ff666-e8fa-4f39-995b-7c16102955a8",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Check that the function performs as we expect:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ed67270-9beb-48df-8c48-93a59ae24254",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert \"training_genes\" in adata_sc.uns\n",
    "assert \"training_genes\" in adata_section.uns\n",
    "\n",
    "print(f\"Number of training_genes: {len(adata_sc.uns['training_genes'])}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59952400-c0aa-49b7-a2fc-c5bfe048b834",
   "metadata": {},
   "source": [
    "#### Computing the map from single-cells to spatial voxels\n",
    "\n",
    "<font size=\"3\"> Having specified the training genes, we can now create the map from dissociated single-cell measurements to the spatial locations. For this, we are going to use the `map_cells_to_space` function from tangram. This function has two different modes, `mode='cells'` and `mode='clusters'`. The latter only maps averaged single-cells which makes the mapping computationally faster and more robust when mapping between specimen."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fb7c48f-11ec-47a8-8836-42c91cec39cb",
   "metadata": {},
   "source": [
    "```python\n",
    "ad_map = tg.map_cells_to_space(\n",
    "    adata_sc,\n",
    "    adata_section,\n",
    "    mode=\"cells\",\n",
    "    density_prior=\"rna_count_based\",\n",
    "    num_epochs=500,\n",
    "    device=\"cpu\",  # or: cpu\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c9b23b9-ff51-48b3-a89a-a8f1945d373a",
   "metadata": {},
   "source": [
    "<font size=\"3\"> I run `tg.map_cells_to_space` for you (takes a while to run), let's just read the resulting AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8849a82d-8176-4cff-93e0-ab07d8e14cda",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_map = ad.read_h5ad(f\"{spatial_wd}/day3/Tangram_data/ad_map_tangram.h5ad\")\n",
    "ad_map"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c9ff6ac-225a-4c63-9f45-76a7b80660d0",
   "metadata": {},
   "source": [
    "<font size=\"3\"> We observe that Tangramâ€™s mapping from cell i to spatial voxels j is stored in the .X property of ad_map.\n",
    "<font size=\"3\">  The meaning of the .var and .obs also changes:\n",
    "\n",
    "- <font size=\"3\"> in .var we have the available metadata of the spatial data, adata_section\n",
    "- <font size=\"3\"> in .obs we have the available metadata of the single-cell data, adata_sc\n",
    "- <font size=\"3\"> In addition, the information about the training run is stored in the .uns key, see `.uns['training_genes_df']` and `.uns[training_history]`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc992f7b-4151-487a-954a-913ca9a2b7a1",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Now we can **project the genes present in the single-cell data to the spatial locations**. This is easily achieved by multiplying the mapping matrix stored in ad_map with the original single-cell data stored in adata_sc. Tangram already provides a convenience function which takes in a mapping and the corresponding single-cell data. The result is a **spatial voxel by genes matrix** which technically is identical to the original spatial data adata_section but **contains expression values for all genes**."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d93c4214-3d7c-4127-8666-6301d7811564",
   "metadata": {},
   "source": [
    "```python\n",
    "ad_ge = tg.project_genes(adata_map=ad_map, adata_sc=adata_sc)\n",
    "ad_ge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d44756c-050f-4630-adc5-209f74745ff8",
   "metadata": {},
   "source": [
    "<font size=\"3\"> I run `tg.project_genes` for you, let's just read the resulting AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62f70844-0288-433e-879d-70438a1ae988",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_ge = ad.read_h5ad(f\"{spatial_wd}/day3/Tangram_data/ad_ge.h5ad\")\n",
    "## add .obsm['X_spatial'] so we can use sc.pl.spatial to plot gene expression\n",
    "ad_ge.obsm['X_spatial'] = np.array(ad_ge.obs[['x','y']])\n",
    "ad_ge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c62d08f-d3ba-4f54-b426-78f2de41f449",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Next, we will compare the new spatial data with the original measurements.\n",
    "\n",
    "<font size=\"3\"> To do this we will plot the true expression and the predicted expression of the shared genes, as well as the Spearman correlation between the true and predicted expression in all the spots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd04efe5-5689-409b-bd10-02df368a2563",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the same set of genes that we predicted with SpaGE\n",
    "# we need to make the genes all lower-case first\n",
    "overlap_genes = [gene.lower() for gene in overlap_genes]\n",
    "overlap_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5af4c94c-9cc9-476b-9602-4d6c023879e8",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Plot the true expression from adata_section"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed36a6d2-750f-42af-ae30-b57791bca2b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(adata_section, color=overlap_genes, spot_size=0.04)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bc740ed-fa95-48e2-aa1f-0ecc22bc94e9",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Plot the predicted expression from ad_ge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75ac35b9-d0e5-4097-a22f-3b40164cf025",
   "metadata": {},
   "outputs": [],
   "source": [
    "ad_ge.var.reset_index(inplace=True)\n",
    "ad_ge.var.set_index('_index', inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ddf2247b-bdaf-428b-8b18-bfc2d54f77b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(ad_ge, color=overlap_genes, spot_size=0.04)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a56f93c8-7f87-440a-bdd2-5666c295012d",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Calculate Spearman correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55bdb832-b668-46ff-8f2a-fe7b87bb0325",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset the AnnData objects to include only the genes in overlap_genes\n",
    "# Assuming both AnnData objects have the same gene ordering\n",
    "ad_true  = adata_section[:, adata_section.var_names.isin(overlap_genes)]\n",
    "ad_ge_predicted = ad_ge[:, ad_ge.var_names.isin(overlap_genes)]\n",
    "\n",
    "# Initialize an empty dictionary to store correlations\n",
    "correlation_dict = {}\n",
    "\n",
    "# Loop over each gene to calculate Spearman correlation\n",
    "for gene in overlap_genes:\n",
    "    # Extract gene expression vectors for the current gene\n",
    "    expr1 = ad_true[:, gene].X.flatten()\n",
    "    expr2 = ad_ge_predicted[:, gene].X.flatten()\n",
    "    \n",
    "    # Calculate Spearman correlation for the current gene\n",
    "    corr, _ = st.spearmanr(expr1, expr2)\n",
    "    \n",
    "    # Store the result in the dictionary\n",
    "    correlation_dict[gene] = corr\n",
    "\n",
    "# Convert the dictionary to a DataFrame\n",
    "tangram_correlations = pd.DataFrame.from_dict(correlation_dict, orient='index', columns=['correlations_tangram'])\n",
    "# capitalize gene names! \n",
    "tangram_correlations.index = tangram_correlations.index.str.capitalize()\n",
    "tangram_correlations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2b5b88e-79b8-467d-bd59-bed051161aaf",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Since we predicted the same genes as SpaGE, let's compare the Spearman correlations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3def41c2-1a5e-48a6-95af-9a97bb26ddd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_correlations = pd.concat([correlations_spage, tangram_correlations],axis=1)\n",
    "all_correlations[['correlations_spage','correlations_tangram']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94b77bb9-ea35-4e9b-b80e-84c88bdb9702",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **Which tool results in higher correlation between the true-measured gene expression and the predicted expression of the genes of interest?**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db86a1c6-50f9-4762-9214-50e9df47082e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Record the end time\n",
    "end_time = time.time()\n",
    "\n",
    "# Calculate elapsed time in minutes\n",
    "elapsed_time = (end_time - start_time) / 60\n",
    "print(f\"Execution Time Tangram: {elapsed_time:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d1582ca-cfd9-461b-98ef-1ec0518cbf61",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Let's delete some objects we won't be using anymore to free some RAM:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ad5b9bb-74e1-40d6-9114-6f89aa370a97",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_sc, ad_map, ad_ge"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ed24aa5-edae-41b7-9cc1-9e14aacfff16",
   "metadata": {},
   "source": [
    "# Spatially-informed clustering with Banksy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abd56a21-fd32-47aa-a9c8-839155fc3c4f",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Spatial omics data are clustered to define both cell types and tissue domains. [BANKSY](https://www.nature.com/articles/s41588-024-01664-3) is an algorithm that unifies these two spatial clustering problems by embedding cells in a product space of their own and the local neighborhood transcriptome, representing cell state and microenvironment, respectively.\n",
    "\n",
    "\n",
    "<figure>\n",
    "    <img src=\"img/banksy.png\" width=\"800\" height=\"300\" alt=\"Description of the image\">\n",
    "    <figcaption>From <a href=\"https://www.nature.com/articles/s41588-024-01664-3\" target=\"_blank\">Singhal et al. (2024) Nature Genetics</a></figcaption>\n",
    "</figure>\n",
    "\n",
    "<font size=\"3\"> This is done by augmenting the  **original gene-cell expression matrix (purple)** **with neighborhood-averaged expression matrices** corresponding to the mean local expression (dark pink) and the AGF (light pink). Here, Î» is a mixing parameter that controls the importance of cellsâ€™ own expression and neighborhood expression effects, G(r) is a radially symmetric Gaussian kernel that decays from magnitude 1 at distanceâ€‰=â€‰0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be33e33c-ec5a-4758-8a08-6873ec738c9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "start_time = time.time()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cd4e2fe-4f62-4854-80c7-742f1ec1d29a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from banksy_utils.load_data import load_adata, display_adata\n",
    "\n",
    "from banksy_utils.filter_utils import normalize_total, filter_hvg, print_max_min\n",
    "\n",
    "# Normalizes the AnnData object\n",
    "adata_section = normalize_total(adata_section)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d59715ef-ab08-40b4-9ac0-42cab1a8fc3d",
   "metadata": {},
   "source": [
    "### Generate spatial weights graph\n",
    "\n",
    "#### In BANKSY, we imagine edges / connections in the graph to represent neighbour relationships between cells\n",
    "\n",
    "<font size=\"3\"> BANKSY requires the following specifications in the main BANKSY algorithm:\n",
    "\n",
    "<font size=\"3\"> 1. The number of spatial neighbours `num_neighbours` (`k_geom`) parameter.\n",
    "    \n",
    "<font size=\"3\"> 2. Assigning weights (dependent on inter-cell spatial distances) to edges in the connected spatial graph. By default, we use the `gaussian decay` option, where weights decay as a function of distance to the index cell with $\\sigma$ = `sigma`.\n",
    "\n",
    "<font size=\"3\"> 3. The Azumithal Gabor Filter parameter `max_m` which indicates whether to use the AGF (`max_m = 1`) or just the mean expression (`max_m = 0`). By default, we set `max_m = 1`.\n",
    "    \n",
    "    \n",
    "#### Construction of the spatial $k_{geom}$ Nearest-Neighbour graph\n",
    "\n",
    "<font size=\"3\"> We represent connections between cells and its neighbours in a graph $G = \\{N,E,W\\}$, comprising of a set of nodes $n \\in N$. Edges represent connectivity between cells $e \\in E$, the edges are weighted $w \\in W$ as a function of the spatial distance between cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dd92fa9-71a6-4267-8878-a7e393730e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "from banksy.main import median_dist_to_nearest_neighbour\n",
    "\n",
    "coord_keys = ('x', 'y', 'X_spatial')\n",
    "\n",
    "# set parameters \n",
    "plot_graph_weights = True\n",
    "k_geom = 15 # number of neighbors\n",
    "max_m = 1 # azumithal transform up to kth order\n",
    "nbr_weight_decay = \"scaled_gaussian\" # can also be \"reciprocal\", \"uniform\" or \"ranked\"\n",
    "\n",
    "# Find median distance to closest neighbours, the median distance will be `sigma`\n",
    "nbrs = median_dist_to_nearest_neighbour(adata_section, key = coord_keys[2])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5b07857-3b88-4bd6-9a55-0c6a79852bcf",
   "metadata": {},
   "source": [
    "#### Generate spatial weights from distance\n",
    "\n",
    "<font size=\"3\"> Here, we generate the spatial weights using the gaussian decay function from the median distance to the k-th nearest neighbours as specified earlier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "354be93b-a5d2-4723-966b-2294c9904922",
   "metadata": {},
   "outputs": [],
   "source": [
    "from banksy.initialize_banksy import initialize_banksy\n",
    "\n",
    "plt.style.use('default')\n",
    "\n",
    "banksy_dict = initialize_banksy(\n",
    "    adata_section,\n",
    "    coord_keys,\n",
    "    k_geom,\n",
    "    nbr_weight_decay=nbr_weight_decay,\n",
    "    max_m=max_m,\n",
    "    plt_edge_hist=True,\n",
    "    plt_nbr_weights=True,\n",
    "    plt_agf_angles=False, # takes long time to plot\n",
    "    plt_theta=True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee047eff-ea5c-457f-a68d-634f6cb88f28",
   "metadata": {},
   "source": [
    "#### Generate BANKSY matrix\n",
    "\n",
    "<font size=\"3\">  Î» is a mixing parameter that controls the importance of cellsâ€™ own expression and neighborhood expression effects, it takes values from 0, being spatial information not used in the clustering, to 1, giving the maximum importance to the neighborhood expression.\n",
    "\n",
    "<font size=\"3\"> To generate the BANKSY matrix, we proceed with the following:\n",
    "\n",
    "<font size=\"3\">1. Matrix multiply sparse CSR weights matrix with cell-gene matrix to get **neighbour matrix** and the **AGF matrix** if `max_m` > 1\n",
    "\n",
    "<font size=\"3\">2. Z-score both matrices along **genes**\n",
    "\n",
    "<font size=\"3\">3. Multiply each matrix by a weighting factor $\\lambda$ (We refer to this parameter as `lambda` in our manuscript and code)\n",
    "\n",
    "<font size=\"3\">4. Concatenate the matrices along the genes dimension in the form -> `horizontal_concat(cell_mat, nbr_mat, agf_mat)`\n",
    "\n",
    "<font size=\"3\">Here, we save all the results in the dictionary (`banksy_dict`), which contains the results from the subsequent operations for BANKSY. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e7f19f4-5423-4501-b021-e12b461a7cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from banksy.embed_banksy import generate_banksy_matrix\n",
    "\n",
    "# The following are the main hyperparameters for BANKSY\n",
    "resolutions = [0.1] # clustering resolution for UMAP\n",
    "pca_dims = [20] # Dimensionality in which PCA reduces to\n",
    "lambda_list = [0.5,0.75] # list of lambda parameters\n",
    "\n",
    "banksy_dict, banksy_matrix = generate_banksy_matrix(adata_section, banksy_dict, lambda_list, max_m)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e49ad78c-89e7-4588-85c3-85b001d3f191",
   "metadata": {},
   "source": [
    "<font size=\"3\"> Append Non-spatial results to the `banksy_dict` for comparsion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f669a59c-d92a-4305-8636-2cf4071f869f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from banksy.main import concatenate_all\n",
    "\n",
    "banksy_dict[\"nonspatial\"] = {\n",
    "    # Here we simply append the nonspatial matrix (adata.X) to obtain the nonspatial clustering results\n",
    "    0.0: {\"adata\": concatenate_all([adata_section.X], 0, adata=adata_section), }\n",
    "}\n",
    "\n",
    "print(banksy_dict['nonspatial'][0.0]['adata'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92ee4953-b93b-4033-8ad4-70a3d8a31567",
   "metadata": {},
   "source": [
    "#### Reduce dimensions of each data matrix\n",
    "<font size=\"3\"> We utilize two common methods for dimensionality reduction:\n",
    "\n",
    "<font size=\"3\"> 1. PCA (using `scikit-learn`), we reduce the size of the matrix from $3*N_{genes}$ to `pca_dims`. As a default settings, we reduce to 20 dimensions.\n",
    "\n",
    "<font size=\"3\"> 2. UMAP (`UMAP` package), which we use to visualize expressions of clusters in the umap space (2-D space)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c645edcc-7748-426b-a22a-795f4ef71497",
   "metadata": {},
   "outputs": [],
   "source": [
    "from banksy_utils.umap_pca import pca_umap\n",
    "\n",
    "pca_umap(banksy_dict,\n",
    "         pca_dims = pca_dims,\n",
    "         add_umap = True,\n",
    "         plt_remaining_var = False,\n",
    "         )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38e23a66-bc17-4a3f-a2cc-993ec6ab1d8d",
   "metadata": {},
   "source": [
    "#### Cluster cells using a partition algorithm\n",
    "\n",
    "<font size=\"3\"> We then cluster cells using the `leiden` algorithm partition methods. Other clustering algorithms include `louvain` (another resolution based clustering algorithm), or `mclust` (a clustering based on gaussian mixture model). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80f93f07-cde9-476f-b86e-a419bbecfe09",
   "metadata": {},
   "outputs": [],
   "source": [
    "from banksy.cluster_methods import run_Leiden_partition\n",
    "seed = 0\n",
    "results_df, max_num_labels = run_Leiden_partition(\n",
    "    banksy_dict,\n",
    "    resolutions,\n",
    "    num_nn = 50,\n",
    "    num_iterations = -1,\n",
    "    partition_seed = seed,\n",
    "    match_labels = True,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3fa2e7d-165c-4c4f-aba1-9e5d39e4ba5c",
   "metadata": {},
   "source": [
    "### Plot results\n",
    "\n",
    "#### Visualize the clustering results from BANKSY, including the clusters from the UMAP embbedings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4671544c-b9e5-48db-b4ba-a31a94e01647",
   "metadata": {},
   "outputs": [],
   "source": [
    "from banksy.plot_banksy import plot_results\n",
    "\n",
    "c_map =  'tab20' # specify color map\n",
    "weights_graph =  banksy_dict['scaled_gaussian']['weights'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc47956f-d639-41c1-ad57-f456affdad68",
   "metadata": {},
   "outputs": [],
   "source": [
    "banksy_path = f'./outputs/banksy_output/' \n",
    "\n",
    "plot_results(\n",
    "    results_df,\n",
    "    weights_graph,\n",
    "    c_map,\n",
    "    match_labels = True,\n",
    "    coord_keys = coord_keys,\n",
    "    max_num_labels  =  max_num_labels, \n",
    "    save_path = os.path.join(banksy_path, 'tmp_png'),\n",
    "    save_fig = True, # save the spatial map of all clusters\n",
    "    save_seperate_fig = True, # save the figure of all clusters plotted seperately\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c510a922-76ac-47cb-82a7-503965e3afac",
   "metadata": {},
   "outputs": [],
   "source": [
    "#  Record the end time\n",
    "end_time = time.time()\n",
    "\n",
    "# Calculate elapsed time in minutes\n",
    "elapsed_time = (end_time - start_time) / 60\n",
    "print(f\"Execution Time Banksy: {elapsed_time:.2f} minutes\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cba5fcf-757f-4174-8332-fe4cf7f0175e",
   "metadata": {},
   "source": [
    "### Investigate Cell Type composition in each Banksy-defined Spatial Domain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f06d763-054e-4fb3-9c11-416ab2950472",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df_lambda05 = results_df.loc['scaled_gaussian_pc20_nc0.50_r0.10','adata']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a1c7f73-9006-44ea-beab-b4ae7f5ded9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df_lambda05.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a7017bf-fd2f-41b7-a26c-7c5d53cd9782",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df_lambda05.obs['Count'] = 1\n",
    "\n",
    "pivot_df = results_df_lambda05.obs.pivot_table(\n",
    "    index='labels_scaled_gaussian_pc20_nc0.50_r0.10', \n",
    "    columns='class', \n",
    "    values='Count', \n",
    "    aggfunc='sum', \n",
    "    fill_value=0\n",
    ")\n",
    "\n",
    "# Step 3: Convert counts to percentages for each SD\n",
    "pivot_df = pivot_df.div(pivot_df.sum(axis=1), axis=0) * 100\n",
    "\n",
    "# Step 4: Set up the plot\n",
    "fig, ax = plt.subplots(figsize=(10, 6))\n",
    "\n",
    "# Plot stacked bars\n",
    "bottom = None\n",
    "for cell_type in pivot_df.columns:\n",
    "    ax.bar(pivot_df.index, pivot_df[cell_type], label=cell_type, bottom=bottom)\n",
    "    bottom = pivot_df[cell_type] if bottom is None else bottom + pivot_df[cell_type]\n",
    "\n",
    "# Step 5: Add labels and title\n",
    "ax.set_xlabel('SD')\n",
    "ax.set_ylabel('Cell Type Composition (%)')\n",
    "ax.set_title('SD vs. Cell Type Composition')\n",
    "ax.set_ylim(0, 100)\n",
    "plt.legend(title=\"Cell Type\", bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "plt.tight_layout()\n",
    "\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fe10078-bf10-4c51-b0a2-b155cfb97790",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 11"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55410926-c22f-425c-92c2-7a0033a0036f",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **What do you expect would happen when we increase `k_geom`?**\n",
    "\n",
    "<span style=\"color: red; font-size: 15px;\"> **Re run Banksy changing `k_geom` to 80.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b09f9e3c-990b-471d-9fcf-328209c6c9ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 12"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1633b929-c7c8-4abd-a5f1-38e08c048e80",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **What do you expect would happen when we set `lambda` = 1?**\n",
    "\n",
    "<span style=\"color: red; font-size: 15px;\"> **How do you expect the spatial clusters plot to look like?**\n",
    "\n",
    "<span style=\"color: red; font-size: 15px;\"> **Re run Banksy changing `lambda` to 1.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e83bbe7-c47f-4702-aef0-11c4055a64f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 13"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffa79a3d-e122-4beb-8c4b-42d9bfd84bac",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **How do the Cell Type compositions of each Spatial Domains change as lambda is increasing? Set lambda=1**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2227c888-f899-4155-a3b9-6436f3437afa",
   "metadata": {},
   "outputs": [],
   "source": [
    "## EX 14"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90143040-bacc-4276-80ce-f3309ae316f8",
   "metadata": {},
   "source": [
    "<span style=\"color: red; font-size: 15px;\"> **How do the Cell Type compositions of each Spatial Domains change as lambda is increasing? Set lambda=0**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa4f46f5-54ab-4a38-a723-6d86396a97b2",
   "metadata": {},
   "source": [
    "### Additional: access image in the AnnData object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "308fd6bd-2050-4773-be52-26260f8280f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from PIL import Image\n",
    "import PIL\n",
    "PIL.Image.MAX_IMAGE_PIXELS = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "291a0ceb-87a0-4fd5-91ee-03bd28e34096",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.datasets.visium_sge(sample_id=\"V1_Mouse_Brain_Sagittal_Posterior\")\n",
    "adata.var_names_make_unique()\n",
    "adata.var[\"mt\"] = adata.var_names.str.startswith(\"MT-\")\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=[\"mt\"], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b3923c1-5836-48c0-bd42-f9946197308f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.filter_cells(adata, min_counts=5000)\n",
    "sc.pp.filter_cells(adata, max_counts=35000)\n",
    "adata = adata[adata.obs[\"pct_counts_mt\"] < 20].copy()\n",
    "print(f\"#cells after MT filter: {adata.n_obs}\")\n",
    "sc.pp.filter_genes(adata, min_cells=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "027519f2-4ec3-444b-a73c-b258ab669bd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "073b9aee-547d-4dd2-9842-2d94c6625c2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Number of spots in the GE matrix (section1, IN TISSUE): ', adata.shape[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba875ae6-60c6-4f20-bccc-6018d672198c",
   "metadata": {},
   "source": [
    "<figure>\n",
    "    <img src=\"img/tissue_hires_image.png\" width=\"700\" height=\"280\" alt=\"Description of the image\">\n",
    "    <figcaption>From <a href=\"https://www.10xgenomics.com/datasets/mouse-brain-serial-section-1-sagittal-posterior-1-standard-1-0-0\" target=\"_blank\">10X Genomics</a></figcaption>\n",
    "</figure>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38fc7f0c-d5fd-4914-ae93-27929db23bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (8, 8)\n",
    "sc.pl.spatial(adata, img_key=\"hires\", color=[\"total_counts\", \"n_genes_by_counts\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ed59d35-ed53-4633-bb39-54bb4391ae4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.pca(adata)\n",
    "sc.pp.neighbors(adata)\n",
    "sc.tl.umap(adata)\n",
    "sc.tl.leiden(adata, key_added=\"clusters\", flavor=\"igraph\", directed=False, n_iterations=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70bb1ae9-0d24-4a96-88c6-5e36e92e9131",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.rcParams[\"figure.figsize\"] = (8, 8)\n",
    "sc.pl.spatial(adata, img_key=\"hires\", color=[\"clusters\"]) # leiden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98c7082a-cd5b-49df-9499-e5d57dfaa4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cluster marker genes\n",
    "adata.var_names_make_unique()\n",
    "sc.tl.rank_genes_groups(adata, \"clusters\", method=\"t-test\")\n",
    "sc.pl.rank_genes_groups_heatmap(adata, groups=\"9\", n_genes=10, groupby=\"clusters\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4c0e5fc-9415-4ce6-9776-def44dd54ecc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pl.spatial(adata, img_key=\"hires\", color=[\"Ddn\"])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:banksy]",
   "language": "python",
   "name": "conda-env-banksy-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
